File: capstoneproject/messaging_system/models/reaction.py
"""
models/reaction.py
Defines the Reaction data model for message interactions.
"""

class Reaction:
    """Represents a user reaction to a message."""
    
    def __init__(self, id: int, message_id: int, user_id: int, reaction_type: str, timestamp: str):
        """Initialize a new reaction instance.
        
        Args:
            id: Unique identifier for the reaction
            message_id: ID of the message being reacted to
            user_id: ID of the user making the reaction
            reaction_type: Type of reaction (like, dislike, etc.)
            timestamp: When the reaction was created
        """
        pass

    def to_dict(self) -> dict:
        """Return a dictionary representation of the reaction.
        
        Returns:
            Dictionary containing reaction data
        """
        pass

    def is_valid_type(self) -> bool:
        """Validate if the reaction type is allowed.
        
        Returns:
            True if reaction type is valid, False otherwise
        """
        pass

File: capstoneproject/app/message/models/Reaction.ts
/**
 * models/Reaction.ts
 * Defines the Reaction data model for message reactions.
 */

export interface ReactionData {
  id: number;
  userId: number;
  messageId: number;
  type: ReactionType;
  timestamp: string;
}

export enum ReactionType {
  LIKE = 'like',
  DISLIKE = 'dislike'
}

export class Reaction {
  /**
   * Represents a user's reaction to a message.
   */
  public id: number;
  public userId: number;
  public messageId: number;
  public type: ReactionType;
  public timestamp: string;

  constructor(data: ReactionData) {
    this.id = data.id;
    this.userId = data.userId;
    this.messageId = data.messageId;
    this.type = data.type;
    this.timestamp = data.timestamp;
  }

  /**
   * Return a plain object representation of the reaction.
   */
  toObject(): ReactionData {
    return {
      id: this.id,
      userId: this.userId,
      messageId: this.messageId,
      type: this.type,
      timestamp: this.timestamp
    };
  }

  /**
   * Check if the reaction is valid.
   */
  isValid(): boolean {
    return Object.values(ReactionType).includes(this.type);
  }

  /**
   * Get emoji representation of the reaction
   */
  getEmoji(): string {
    const emojiMap: Record<ReactionType, string> = {
      [ReactionType.LIKE]: '👍',
      [ReactionType.DISLIKE]: '👎'
    };
    return emojiMap[this.type] || '👍';
  }

  /**
   * Check if the reaction type is valid (only like/dislike allowed)
   */
  isValidType(): boolean {
    return ['like', 'dislike'].includes(this.type);
  }
}
